doc/util/collect-syntax.py

```diff
@@ -0,0 +1,327 @@
+        result.append(clean)
+        continue
+    if ok:
+      result.append(ok)
+  return result
+
+  
+def get_nonterminals_in_expansion(ident, stack, applied):
+  global _productions
+  fname, lineno, blob = _productions[ident]
+  vprint('Inspecting expansions of:', ident)
+  vprint(blob)
+  nonterminals = []
+  for line in blob.split('\n'):
+    if re.match(_match_production, line) or is_meta_sentence(line):
+      continue
+    nonterminals += get_nonterminals(line)
+  vprint('Found', len(nonterminals), 'nonterminals')
+  for n in nonterminals:
+    if is_meta_keyword(n):
+      continue
+    if not n in _productions:
+      msg = 'Nonterminal not in production list: ' + str(n)
+      vprint(msg)
+      raise Exception(msg)
+    if not n in applied:
+      stack.append(n)
+      applied.add(n)
+  return
+
+
+def output_productions_application_order(out):
+  out.write('.. _Productions_Listed_Application_Order:\n')
+  out.write('\n')
+  title = 'Productions Listed in Application Order'
+  out.write(title + '\n')
+  out.write((len(title) * '-') + '\n')
+  out.write('\n')
+  out.write('.. code-block:: syntax\n')
+  out.write('\n')
+  global _productions
+  start = 'module-declaration-statement'
+  vprint('Computing application-first ordering starting with:', start)
+  stack = [start]
+  applied = {start}
+  # Do a breadth-first search over all productions...
+  while stack:
+    ident = stack.pop(0)
+    get_nonterminals_in_expansion(ident, stack, applied)
+    _, _, blob = _productions[ident]
+    out.write(blob)
+    out.write('\n')
+  # Consider unapplied productions to be a fatal error...
+  unapplied_count = len(_productions) - len(applied)
+  if unapplied_count != 0:
+    msg = str(unapplied_count) + ' nonterminals are never applied'
+    vprint(msg)
+    for ident in _productions.keys():
+      if not ident in applied:
+        vprint(ident)
+    raise Exception(msg)
+  return
+
+
+def main(args):
+  global _do_list_productions, _rstfiles, _outdir, _verbose
+  _do_list_productions = args['list_productions']
+  _rstfiles = args['rstfiles']
+  _outdir = args['outdir']
+  _verbose = args['verbose']
+  collect_productions_from_files()
+  if _do_list_productions:
+    global _productions
+    print('Collected', len(_productions))
+    for ident in sorted(_productions.keys()):
+      fname, lineno, blob = _productions[ident]
+      print(ident, 'in', fname, 'on line', lineno)
+      print(blob)
+  outname = os.path.join(_outdir, 'syntax.rst')
+  vprint('Writing output to:', outname)  
+  with open(outname, 'w') as outfile:
+    output_prelude(outfile)
+    output_productions_alphabetical(outfile)
+    output_productions_application_order(outfile)
+  return
+
+
+if __name__ == '__main__':
+  err = 0
+  try:
+    # Parse arguments and cast them into a dictionary.
+    arguments = vars(get_script_arguments())
+    main(arguments)
+  except Exception as e:
+    traceback.print_exc()
+    err = 1
+  sys.exit(err)
```

Comment:
I am susprised you need this try/except/exit. Doesn't all that happen by default?

Reply:
I wasn't sure, I just wanted to make it so you could collect the exit code, but perhaps not necessary :).